File: ios/MimiRemote/Sources/Core/API/AgentAPIClient.swift
import Foundation

enum AgentAPIError: LocalizedError {
    case invalidEndpoint
    case invalidResponse
    case server(status: Int, message: String)
    case decoding(Error)

    var errorDescription: String? {
        switch self {
        case .invalidEndpoint:
            return "Endpoint 无效"
        case .invalidResponse:
            return "agentd 返回了无效响应"
        case .server(let status, let message):
            return "HTTP \(status)：\(message)"
        case .decoding(let error):
            return "解析响应失败：\(error.localizedDescription)"
        }
    }
}

struct AgentAPIClient {
    let endpoint: String
    let token: String

    private var baseURL: URL? {
        URL(string: Self.normalizedEndpoint(endpoint))
    }

    func health() async throws -> HealthResponse {
        try await request(path: "/healthz", method: "GET", requiresAuth: false, body: Optional<Data>.none)
    }

    func claimPairing(_ claim: PairingClaimRequest) async throws -> PairingClaimResponse {
        let body = try JSONEncoder().encode(claim)
        return try await request(path: "/api/pair/claim", method: "POST", requiresAuth: false, body: body)
    }

    func version() async throws -> VersionResponse {
        try await request(path: "/api/version", method: "GET", body: Optional<Data>.none)
    }

    func appServerConfig() async throws -> CodexAppServerConfigResponse {
        try await request(path: "/api/app-server/config", method: "GET", body: Optional<Data>.none)
    }

    func relayDiagnostics() async throws -> RelayDiagnosticsResponse {
        try await request(path: "/api/diagnostics/relay", method: "GET", body: Optional<Data>.none)
    }

    func capabilities(path: String?) async throws -> CapabilityListResponse {
        let body = try JSONEncoder().encode(CapabilityListRequest(path: path))
        return try await request(path: "/api/capabilities/list", method: "POST", body: body)
    }

    func projects() async throws -> [AgentProject] {
        let response: ProjectsResponse = try await request(path: "/api/projects", method: "GET", body: Optional<Data>.none)
        return response.projects
    }

    func resolveWorkspace(path: String) async throws -> AgentWorkspace {
        let body = try JSONEncoder().encode(WorkspaceResolveRequest(path: path))
        let response: WorkspaceResolveResponse = try await request(path: "/api/workspaces/resolve", method: "POST", body: body)
        return response.workspace
    }

    func createWorktree(path: String, name: String?, base: String?, branch: String?) async throws -> WorktreeCreateResponse {
        let body = try JSONEncoder().encode(WorktreeCreateRequest(path: path, name: name, base: base, branch: branch))
        return try await request(path: "/api/worktrees/create", method: "POST", body: body)
    }

    func worktreeBranches(path: String) async throws -> WorktreeBranchListResponse {
        let body = try JSONEncoder().encode(WorktreeBranchListRequest(path: path))
        return try await request(path: "/api/worktrees/branches", method: "POST", body: body)
    }

    func listWorktrees() async throws -> [WorktreeListItem] {
        let response: WorktreeListResponse = try await request(path: "/api/worktrees/list", method: "GET", body: Optional<Data>.none)
        return response.worktrees
    }

    func deleteWorktree(path: String, force: Bool = false) async throws -> WorktreeDeleteResponse {
        let body = try JSONEncoder().encode(WorktreeDeleteRequest(path: path, force: force))
        return try await request(path: "/api/worktrees/delete", method: "POST", body: body)
    }

    func pruneMissingWorktrees() async throws -> WorktreePruneResponse {
        try await request(path: "/api/worktrees/prune", method: "POST", body: Optional<Data>.none)
    }

    func listDirectories(path: String) async throws -> DirectoryListResponse {
        let body = try JSONEncoder().encode(DirectoryListRequest(path: path))
        return try await request(path: "/api/directories/list", method: "POST", body: body)
    }

    func readFile(path: String) async throws -> FileReadResponse {
        let body = try JSONEncoder().encode(FileReadRequest(path: path))
        return try await request(path: "/api/files/read", method: "POST", body: body)
    }

    func commandActions(path: String) async throws -> [AgentCommandAction] {
        let body = try JSONEncoder().encode(CommandActionListRequest(path: path))
        let response: CommandActionListResponse = try await request(path: "/api/actions/list", method: "POST", body: body)
        return response.actions
    }

    func runCommandAction(path: String, id: String, confirmed: Bool) async throws -> CommandActionRunResponse {
        let body = try JSONEncoder().encode(CommandActionRunRequest(path: path, id: id, confirmed: confirmed))
        return try await request(path: "/api/actions/run", method: "POST", body: body)
    }

    func gitStatus(path: String) async throws -> GitStatusResponse {
        let body = try JSONEncoder().encode(GitStatusRequest(path: path))
        return try await request(path: "/api/git/status", method: "POST", body: body)
    }

    func gitAction(path: String, action: GitActionKind, files: [String]) async throws -> GitStatusResponse {
        let body = try JSONEncoder().encode(GitActionRequest(path: path, action: action, files: files))
        return try await request(path: "/api/git/action", method: "POST", body: body)
    }

    func gitPatchAction(path: String, action: GitActionKind, patch: String) async throws -> GitStatusResponse {
        let body = try JSONEncoder().encode(GitActionRequest(path: path, action: action, patch: patch))
        return try await request(path: "/api/git/action", method: "POST", body: body)
    }

    func gitCommit(path: String, message: String) async throws -> GitStatusResponse {
        let body = try JSONEncoder().encode(GitCommitRequest(path: path, message: message))
        return try await request(path: "/api/git/commit", method: "POST", body: body)
    }

    func gitPush(path: String, remote: String?) async throws -> GitPushResponse {
        let body = try JSONEncoder().encode(GitPushRequest(path: path, remote: remote))
        return try await request(path: "/api/git/push", method: "POST", body: body)
    }

    func gitCreatePullRequest(path: String, title: String, body prBody: String, draft: Bool) async throws -> GitPullRequestResponse {
        let body = try JSONEncoder().encode(GitPullRequestRequest(path: path, title: title, body: prBody, draft: draft))
        return try await request(path: "/api/git/pull-request", method: "POST", body: body)
    }

    func gitPullRequestStatus(path: String) async throws -> GitPullRequestStatusResponse {
        let body = try JSONEncoder().encode(GitPullRequestStatusRequest(path: path))
        return try await request(path: "/api/git/pull-request/status", method: "POST", body: body)
    }

    func transcribeVoice(
        filename: String,
        contentType: String,
        audioData: Data,
        language: String?,
        prompt: String?
    ) async throws -> VoiceTranscriptionResponse {
        let body = try JSONEncoder().encode(VoiceTranscriptionRequest(
            filename: filename,
            contentType: contentType,
            audioBase64: audioData.base64EncodedString(),
            language: language,
            prompt: prompt
        ))
        return try await request(path: "/api/voice/transcribe", method: "POST", body: body, timeout: 60)
    }

    private func request<T: Decodable>(
        path: String,
        method: String,
        requiresAuth: Bool = true,
        body: Data?,
        timeout: TimeInterval = 20
    ) async throws -> T {
        guard let baseURL else {
            throw AgentAPIError.invalidEndpoint
        }
        guard let url = makeURL(baseURL: baseURL, path: path) else {
            throw AgentAPIError.invalidEndpoint
        }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.timeoutInterval = timeout
        if requiresAuth {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let body {
            request.httpBody = body
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw AgentAPIError.invalidResponse
        }
        guard (200..<300).contains(http.statusCode) else {
            throw AgentAPIError.server(status: http.statusCode, message: decodeError(data))
        }
        if T.self == EmptyResponse.self {
            return EmptyResponse() as! T
        }
        do {
            return try Self.decoder.decode(T.self, from: data)
        } catch {
            throw AgentAPIError.decoding(error)
        }
    }

    private func decodeError(_ data: Data) -> String {
        if let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let error = object["error"] as? String {
            return error
        }
        return String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? "未知错误"
    }

    private func makePath(_ path: String, query: [String: String?]) -> String {
        var components = URLComponents()
        components.path = path
        components.queryItems = query.compactMap { key, value in
            guard let value, !value.isEmpty else {
                return nil
            }
            return URLQueryItem(name: key, value: value)
        }
        return components.string ?? path
    }

    private func makeURL(baseURL: URL, path: String) -> URL? {
        guard var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false) else {
            return nil
        }
        let normalizedPath = path.hasPrefix("/") ? path : "/" + path
        guard let pathComponents = URLComponents(string: normalizedPath) else {
            return nil
        }
        components.path = pathComponents.path
        components.queryItems = pathComponents.queryItems
        return components.url
    }

    static func normalizedEndpoint(_ raw: String) -> String {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.hasPrefix("http://") || trimmed.hasPrefix("https://") {
            return trimmed.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        }
        return "http://" + trimmed.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    }

    static let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .custom { decoder in
            let container = try decoder.singleValueContainer()
            let raw = try container.decode(String.self)
            if let date = DateParsers.iso8601Fractional.date(from: raw) {
                return date
            }
            if let date = DateParsers.iso8601.date(from: raw) {
                return date
            }
            throw DecodingError.dataCorruptedError(in: container, debugDescription: "日期格式无效：\(raw)")
        }
        return decoder
    }()
}

private struct EmptyResponse: Decodable {}

private enum DateParsers {
    static let iso8601Fractional: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()

    static let iso8601: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter
    }()
}

File: android/app/src/main/java/com/gaixianggeng/mimi/core/network/AgentApiClient.kt
package com.gaixianggeng.mimi.core.network

import com.gaixianggeng.mimi.core.model.AgentProject
import com.gaixianggeng.mimi.core.model.AgentProjectsResponse
import com.gaixianggeng.mimi.core.model.HealthResponse
import com.gaixianggeng.mimi.core.model.GitActionKind
import com.gaixianggeng.mimi.core.model.GitActionRequest
import com.gaixianggeng.mimi.core.model.GitMutationPolicy
import com.gaixianggeng.mimi.core.model.GitPublishPolicy
import com.gaixianggeng.mimi.core.model.GitStatusPolicy
import com.gaixianggeng.mimi.core.model.GitCommitRequest
import com.gaixianggeng.mimi.core.model.GitPushRequest
import com.gaixianggeng.mimi.core.model.GitPushResponse
import com.gaixianggeng.mimi.core.model.GitStatusRequest
import com.gaixianggeng.mimi.core.model.GitStatusResponse
import com.gaixianggeng.mimi.core.model.PairingClaimRequest
import com.gaixianggeng.mimi.core.model.PairingClaimResponse
import com.gaixianggeng.mimi.core.model.VoiceTranscriptionRequest
import com.gaixianggeng.mimi.core.model.VoiceTranscriptionResponse
import com.gaixianggeng.mimi.core.model.FileReadRequest
import com.gaixianggeng.mimi.core.model.FileReadResponse
import com.gaixianggeng.mimi.core.model.WorktreeListResponse
import com.gaixianggeng.mimi.core.model.WorktreeBranchListRequest
import com.gaixianggeng.mimi.core.model.WorktreeBranchListResponse
import com.gaixianggeng.mimi.core.model.RelayDiagnosticsResponse
import com.gaixianggeng.mimi.core.model.TailscaleNetworkPathResponse
import com.gaixianggeng.mimi.core.model.WorktreeCreateRequest
import com.gaixianggeng.mimi.core.model.WorktreeCreateResponse
import com.gaixianggeng.mimi.core.model.WorktreeDeleteRequest
import com.gaixianggeng.mimi.core.model.WorktreeDeleteResponse
import com.gaixianggeng.mimi.core.model.WorktreePruneResponse
import com.gaixianggeng.mimi.core.model.WorktreeCleanupRequest
import com.gaixianggeng.mimi.core.model.WorktreeCleanupResponse
import com.gaixianggeng.mimi.core.model.DoctorResults
import com.gaixianggeng.mimi.core.model.CapabilityListRequest
import com.gaixianggeng.mimi.core.model.LegacyCapabilityListResponse
import com.gaixianggeng.mimi.core.model.WorkspaceResolveRequest
import com.gaixianggeng.mimi.core.model.WorkspaceResolveResponse
import com.gaixianggeng.mimi.core.model.AgentWorkspace
import com.gaixianggeng.mimi.core.model.DirectoryListRequest
import com.gaixianggeng.mimi.core.model.DirectoryListResponse
import com.gaixianggeng.mimi.core.model.CommandActionListRequest
import com.gaixianggeng.mimi.core.model.CommandActionListResponse
import com.gaixianggeng.mimi.core.model.CommandActionRunRequest
import com.gaixianggeng.mimi.core.model.CommandActionRunResponse
import com.gaixianggeng.mimi.core.model.GitQuickPublishRequest
import com.gaixianggeng.mimi.core.model.GitQuickPublishResponse
import com.gaixianggeng.mimi.core.model.GitPullRequestRequest
import com.gaixianggeng.mimi.core.model.GitPullRequestResponse
import com.gaixianggeng.mimi.core.model.GitPullRequestStatusRequest
import com.gaixianggeng.mimi.core.model.GitPullRequestStatusResponse
import com.gaixianggeng.mimi.core.model.GitTestFlightStatusRequest
import com.gaixianggeng.mimi.core.model.GitTestFlightRunRequest
import com.gaixianggeng.mimi.core.model.GitTestFlightStatusResponse
import com.gaixianggeng.mimi.core.model.AppServerConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.io.IOException
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

class AgentApiException(message: String, val statusCode: Int? = null) : IOException(message)

class AgentApiClient(
    private val client: OkHttpClient,
    private val json: Json,
) {
    suspend fun ready(endpoint: String, token: String): HealthResponse =
        get(endpoint, "/api/readyz", token)

    suspend fun doctor(endpoint: String, token: String): DoctorResults =
        get(endpoint, "/api/doctor", token)

    suspend fun capabilities(endpoint: String, token: String, path: String?): LegacyCapabilityListResponse =
        post(endpoint, "/api/capabilities/list", token, CapabilityListRequest(path))

    suspend fun projects(endpoint: String, token: String): List<AgentProject> =
        get<AgentProjectsResponse>(endpoint, "/api/projects", token).projects

    suspend fun appServerConfig(endpoint: String, token: String): AppServerConfig =
        get(endpoint, "/api/app-server/config", token)

    suspend fun relayDiagnostics(endpoint: String, token: String): RelayDiagnosticsResponse =
        get(endpoint, "/api/diagnostics/relay", token)

    suspend fun tailscaleNetworkPath(endpoint: String, token: String): TailscaleNetworkPathResponse =
        get(endpoint, "/api/diagnostics/tailscale-path", token)

    suspend fun codexHistoryDiagnostics(endpoint: String, token: String, projectId: String, limit: Int = 120): JsonObject {
        val normalizedLimit = limit.coerceIn(1, 200)
        val normalizedProject = projectId.trim()
        require(normalizedProject.length in 1..512) { "A bounded project scope is required" }
        val encodedProject = java.net.URLEncoder.encode(normalizedProject, Charsets.UTF_8.name()).replace("+", "%20")
        val query = buildString {
            append("?limit=").append(normalizedLimit)
            append("&project_id=").append(encodedProject)
        }
        return get(endpoint, "/api/debug/codex-history$query", token)
    }

    suspend fun resolveWorkspace(endpoint: String, token: String, path: String): AgentWorkspace =
        post<WorkspaceResolveRequest, WorkspaceResolveResponse>(endpoint, "/api/workspaces/resolve", token, WorkspaceResolveRequest(path)).workspace

    suspend fun listDirectories(endpoint: String, token: String, path: String): DirectoryListResponse =
        post(endpoint, "/api/directories/list", token, DirectoryListRequest(path))

    suspend fun commandActions(endpoint: String, token: String, path: String): CommandActionListResponse {
        val normalizedPath = boundedActionField(path, 4096, "Action path")
        return post(endpoint, "/api/actions/list", token, CommandActionListRequest(normalizedPath))
    }

    suspend fun runCommandAction(endpoint: String, token: String, path: String, id: String, confirmed: Boolean): CommandActionRunResponse {
        val normalizedPath = boundedActionField(path, 4096, "Action path")
        val normalizedId = boundedActionField(id, 256, "Action id")
        return post(endpoint, "/api/actions/run", token, CommandActionRunRequest(normalizedPath, normalizedId, confirmed))
    }

    private fun boundedActionField(value: String, maxLength: Int, label: String): String {
        val normalized = value.trim()
        require(value == normalized && value.length in 1..maxLength) { "$label is invalid" }
        return value
    }

    suspend fun gitStatus(endpoint: String, token: String, path: String): GitStatusResponse =
        post<GitStatusRequest, GitStatusResponse>(
            endpoint,
            "/api/git/status",
            token,
            GitPublishPolicy.statusRequest(path),
        ).let { GitStatusPolicy.sanitize(it, path) }

    suspend fun gitAction(endpoint: String, token: String, path: String, action: GitActionKind, files: List<String>): GitStatusResponse =
        post<GitActionRequest, GitStatusResponse>(
            endpoint,
            "/api/git/action",
            token,
            GitMutationPolicy.fileRequest(path, action, files),
        ).let { GitStatusPolicy.sanitize(it, path) }

    suspend fun gitPatchAction(endpoint: String, token: String, path: String, action: GitActionKind, patch: String): GitStatusResponse =
        post<GitActionRequest, GitStatusResponse>(
            endpoint,
            "/api/git/action",
            token,
            GitMutationPolicy.patchRequest(path, action, patch),
        ).let { GitStatusPolicy.sanitize(it, path) }

    suspend fun gitCommit(endpoint: String, token: String, path: String, message: String): GitStatusResponse =
        post<GitCommitRequest, GitStatusResponse>(
            endpoint,
            "/api/git/commit",
            token,
            GitPublishPolicy.commitRequest(path, message),
        ).let { GitStatusPolicy.sanitize(it, path) }

    suspend fun gitPush(endpoint: String, token: String, path: String, remote: String? = null): GitPushResponse =
        post<GitPushRequest, GitPushResponse>(endpoint, "/api/git/push", token, GitPublishPolicy.pushRequest(path, remote))
            .let(GitPublishPolicy::bound)
            .let { it.copy(status = GitStatusPolicy.sanitize(it.status, path)) }

    suspend fun gitQuickPublish(endpoint: String, token: String, path: String, message: String, confirmed: Boolean): GitQuickPublishResponse =
        post<GitQuickPublishRequest, GitQuickPublishResponse>(
            endpoint,
            "/api/git/quick-publish",
            token,
            GitPublishPolicy.quickPublishRequest(path, message, confirmed),
        ).let(GitPublishPolicy::bound)
            .let { it.copy(status = GitStatusPolicy.sanitize(it.status, path)) }

    suspend fun gitCreatePullRequest(endpoint: String, token: String, path: String, title: String, body: String, draft: Boolean): GitPullRequestResponse =
        post<GitPullRequestRequest, GitPullRequestResponse>(
            endpoint,
            "/api/git/pull-request",
            token,
            GitPublishPolicy.pullRequest(path, title, body, draft),
        ).let(GitPublishPolicy::bound)

    suspend fun gitPullRequestStatus(endpoint: String, token: String, path: String): GitPullRequestStatusResponse =
        post(endpoint, "/api/git/pull-request/status", token, GitPublishPolicy.pullRequestStatus(path))

    suspend fun gitTestFlightStatus(endpoint: String, token: String, path: String): GitTestFlightStatusResponse =
        post<GitTestFlightStatusRequest, GitTestFlightStatusResponse>(
            endpoint,
            "/api/git/testflight/status",
            token,
            GitPublishPolicy.testFlightStatus(path),
        ).let(GitPublishPolicy::bound)

    suspend fun gitTestFlightRun(endpoint: String, token: String, path: String, whatToTest: String, confirmed: Boolean): GitTestFlightStatusResponse =
        post<GitTestFlightRunRequest, GitTestFlightStatusResponse>(
            endpoint,
            "/api/git/testflight/run",
            token,
            GitPublishPolicy.testFlightRun(path, whatToTest, confirmed),
        ).let(GitPublishPolicy::bound)

    suspend fun transcribeVoice(
        endpoint: String,
        token: String,
        filename: String,
        contentType: String,
        audioBase64: String,
        language: String? = null,
    ): VoiceTranscriptionResponse = post(
        endpoint,
        "/api/voice/transcribe",
        token,
        VoiceTranscriptionRequest(filename, contentType, audioBase64, language),
    )

    suspend fun readFile(endpoint: String, token: String, path: String): FileReadResponse =
        post(endpoint, "/api/files/read", token, FileReadRequest(path))

    suspend fun readHistoryMedia(endpoint: String, token: String, id: String): FileReadResponse {
        val normalized = id.trim()
        require(normalized.isNotEmpty() && normalized.length <= 256) { "History media id is invalid" }
        val encoded = java.net.URLEncoder.encode(normalized, Charsets.UTF_8.name()).replace("+", "%20")
        return get(endpoint, "/api/app-server/history-media/$encoded", token)
    }

    suspend fun listWorktrees(endpoint: String, token: String): WorktreeListResponse =
        get(endpoint, "/api/worktrees/list", token)

    suspend fun listWorktreeBranches(endpoint: String, token: String, path: String): WorktreeBranchListResponse {
        val normalizedPath = boundedWorktreePath(path)
        return post(endpoint, "/api/worktrees/branches", token, WorktreeBranchListRequest(normalizedPath))
    }

    suspend fun createWorktree(endpoint: String, token: String, path: String, name: String?, base: String?): WorktreeCreateResponse {
        val normalizedPath = boundedWorktreePath(path)
        val normalizedName = name?.trim()?.takeIf(String::isNotEmpty)
        val normalizedBase = base?.trim()?.takeIf(String::isNotEmpty)
        require(normalizedName == null || normalizedName.length <= 256) { "Worktree name is too long" }
        require(normalizedBase == null || normalizedBase.length <= 512) { "Worktree base is too long" }
        return post(
            endpoint,
            "/api/worktrees/create",
            token,
            WorktreeCreateRequest(normalizedPath, normalizedName, normalizedBase),
        )
    }

    suspend fun deleteWorktree(endpoint: String, token: String, path: String): WorktreeDeleteResponse =
        post(endpoint, "/api/worktrees/delete", token, WorktreeDeleteRequest(boundedWorktreePath(path), force = false))

    suspend fun pruneWorktrees(endpoint: String, token: String): WorktreePruneResponse =
        post(endpoint, "/api/worktrees/prune", token, buildJsonObject { })

    private fun boundedWorktreePath(path: String): String {
        val normalized = path.trim()
        require(normalized.length in 1..4096) { "Worktree path is invalid" }
        return normalized
    }

    suspend fun previewWorktreeCleanup(endpoint: String, token: String): WorktreeCleanupResponse =
        post(endpoint, "/api/worktrees/cleanup", token, WorktreeCleanupRequest())

    suspend fun executeWorktreeCleanup(endpoint: String, token: String, paths: List<String>, planId: String): WorktreeCleanupResponse {
        require(paths.size in 1..500) { "Cleanup candidates are invalid" }
        require(paths.distinct().size == paths.size) { "Cleanup candidates must be unique" }
        require(paths.all { it == it.trim() && it.length in 1..4096 }) { "Cleanup candidate path is invalid" }
        require(planId == planId.trim() && planId.length in 1..512) { "Cleanup plan id is invalid" }
        return post(endpoint, "/api/worktrees/cleanup", token, WorktreeCleanupRequest(false, true, paths, planId))
    }

    suspend fun claimPairing(endpoint: String, claim: PairingClaimRequest): PairingClaimResponse {
        val normalized = normalizedEndpoint(endpoint)
        val body = json.encodeToString(claim).toRequestBody(JSON_MEDIA_TYPE)
        val request = Request.Builder()
            .url("$normalized/api/pair/claim")
            .post(body)
            .build()
        return executeAndDecode(request)
    }

    private suspend inline fun <reified T> get(endpoint: String, path: String, token: String): T {
        val request = Request.Builder()
            .url("${normalizedEndpoint(endpoint)}$path")
            .header("Authorization", "Bearer $token")
            .get()
            .build()
        return executeAndDecode(request)
    }

    private suspend inline fun <reified RequestType, reified ResponseType> post(
        endpoint: String,
        path: String,
        token: String,
        bodyValue: RequestType,
    ): ResponseType {
        val body = json.encodeToString(bodyValue).toRequestBody(JSON_MEDIA_TYPE)
        val request = Request.Builder()
            .url("${normalizedEndpoint(endpoint)}$path")
            .header("Authorization", "Bearer $token")
            .post(body)
            .build()
        return executeAndDecode(request)
    }

    private fun normalizedEndpoint(raw: String): String = when (val assessment = EndpointPolicy.assess(raw)) {
        is EndpointAssessment.Allowed -> assessment.normalizedEndpoint
        is EndpointAssessment.Rejected -> throw AgentApiException(assessment.reason)
    }

    private suspend inline fun <reified T> executeAndDecode(request: Request): T {
        val response = client.newCall(request).await()
        return withContext(Dispatchers.IO) {
            response.use {
                val body = it.body.string()
                if (!it.isSuccessful) {
                    val safeMessage = when (it.code) {
                        401 -> "Authentication failed"
                        403 -> "Request was forbidden"
                        else -> "Server returned HTTP ${it.code}"
                    }
                    throw AgentApiException(safeMessage, it.code)
                }
                try {
                    json.decodeFromString(body)
                } catch (error: kotlinx.serialization.SerializationException) {
                    throw AgentApiException(
                        "Could not decode ${request.url.encodedPath} response: " +
                            (error.message ?: error::class.simpleName.orEmpty()),
                    )
                }
            }
        }
    }

    private suspend fun Call.await(): Response = suspendCancellableCoroutine { continuation ->
        continuation.invokeOnCancellation { cancel() }
        enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                if (continuation.isActive) continuation.resumeWithException(e)
            }

            override fun onResponse(call: Call, response: Response) {
                if (continuation.isActive) continuation.resume(response) else response.close()
            }
        })
    }

    private companion object {
        val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
    }
}
